train.py

```diff
@@ -2,84 +2,135 @@
+                        scheduler_d.step()
+
+                g_loss = self.generator_loss((batch_size, parameters.gan_noise_size))
 
                 optimizer_g.zero_grad()
-                loss.backward()
+                g_loss.backward()
                 optimizer_g.step()
 
                 if scheduler_g is not None:
                     scheduler_g.step()
 
-            X_batch_real = X_batch_real.to(device)
+                if experiment is not None and iterations_total % parameters.log_every == 0:
+                    assert test_dataset is not None
+                    experiment.log_metrics({'g_loss': g_loss.detach().cpu().numpy(),
+                                            'd_loss': d_loss.detach().cpu().numpy()})
+                    eval_batch_num = int((parameters.gan_test_ratio * len(test_dataset)) / parameters.eval_batch_size)
+                    evaluate_model(self.generator, experiment,
+                                   test_dataset, parameters.eval_batch_size,
+                                   eval_batch_num, parameters,
+                                   self.device, scaler, iterations_total)
+
+        return iterations_total
+
+    def discriminator_loss(self, parameters, X_batch_real):
+        X_batch_real = X_batch_real.to(self.device)
+
+        X_noise = torch.randn((X_batch_real.size(0), parameters.gan_noise_size)).to(self.device)
+        G_output = self.generator(X_noise)
+
+        D_fake = self.discriminator(G_output.float())
+        D_real = self.discriminator(X_batch_real.float())
+
+        D_loss = -torch.mean(torch.cat((torch.log(D_real + 1e-8),
+                                        torch.log(1 - D_fake + 1e-8))))
+
+        return D_loss
+
+    def generator_loss(self, noise_batch_shape):
+        X_noise = torch.randn(noise_batch_shape).to(self.device)
+        G_output = self.generator(X_noise)
+        D_output = self.discriminator(G_output.float())
+
+        G_loss = -torch.mean(torch.log(D_output + 1e-8))
+
+        return G_loss
+
+
+class WGPGANTrainer(GANTrainer):
+    def discriminator_loss(self, parameters, X_batch_real):
+        X_batch_real = X_batch_real.to(self.device)
+
+        X_noise = torch.randn((parameters.batch_size, parameters.gan_noise_size)).to(self.device)
+        G_output = self.generator(X_noise).detach()
+
+        D_fake = self.discriminator(G_output.float())
+        D_real = self.discriminator(X_batch_real.float())
 
-            X_noise = torch.randn((batch_size, parameters.gan_noise_size), device=device)
-            X_batch_fake = generator(X_noise).detach()
+        epsilon = torch.rand(X_batch_real.shape[0], 1).expand(X_batch_real.size()).to(self.device)
 
-            predict = discriminator(X_batch_real.float())
-            loss = criterion(predict, y_real)
+        G_interpolation = epsilon * X_batch_real.float() + (1 - epsilon) * G_output.float()
+        G_interpolation = torch.autograd.Variable(G_interpolation, requires_grad=True)
```

Comment:
This is redundant. Deprecated in torch scince 0.4.0 https://pytorch.org/blog/pytorch-0_4_0-migration-guide/

Reply:
Fixed.